### ISSUE TITLE
Getting matrix functionality to work

### ISSUE THREAD

**Issue by trych** [basiljs/basil.js#222]
## The issue
(I try to describe my understandig of the issue we currently have, please someone edit it to add some missing stuff or to correct me)

Currently we have transformation methods such as `b.translate()`, `b.rotate()` and `b.scale()`. These are – in theory – supposed to mimic the behavior of Processing's equivalents `translate()`, `rotate()` and `scale()`.

In Processing all these methods relate to the origin of the scene (by default [0, 0], _top_ left corner), and they get applied to the shapes that will be added after calling the methods. So, when I do something like

```js
rotate( radians(45) );
rect(50, 0, 20, 20);
```
the rect is not rotated by `45°` around _itself_, but around the origin.

Here's where basil is different at the moment. If I do

```js
b.rotate( b.radians(45) );
b.rect(50, 0, 20, 20);
```

the rect is created at position `[0, 50]` and then rotated by 45 around _itself_ (or rather around its own origin, which is whatever InDesign is currently set to, most probably the default top left corner, but could be any other corner or the center).

`b.scale()` has the same problem that it only relates to the shape's own origin, not to the scene origin. `b.translate()` seems to behave as intended (?), so if I do: 

```js
b.translate(20, 30);
b.rect(0, 0, 50, 50);
```
the rect _does_ get created at `[20, 30]`.

Additionally there are a bunch of methods `b.Matrix2D`, `b.Matrix2D.rotate()` that are supposedly transform the Adobe matrix, but currently nobody really seems to know how to use them or if they ever worked (please correct me, if I am wrong).

## Proposed Solution
The proposal would be to make all three transformations method act like they do in Processing, ideally utilizing Adobe's matrix system and to figure out if (and why) we need separate `b.Matrix2D` methods.

(For transformations around the shape's own origins there will be a new `b.transform(shape, "transformationName", value)` method (that is already taken care of)).

## Resources
[Coordinate Spaces & Transformations in InDesign — Chap.1-3](http://www.indiscripts.com/post/2014/03/coordinate-spaces-and-transformations-1)

[Coordinate Spaces & Transformations in InDesign — Chap.4](http://www.indiscripts.com/post/2016/06/coordinate-spaces-and-transformations-4)

Most certainly these documents give the best overview over the topic, as (quote Marc Autret) _"the fault [of being an obscure topic] mainly lies with Adobe documentation, especially the Scripting DOM reference, which does not clearly explain the topic and some of its essential keys"_.

@bohnacker Thanks in advance for having a look at this. :)

**Comment by ffd8** [comment]
Sorry for the delayed chime in on this issue.. indeed – this is why that nice `b.transform(obj, "transformationProp", value)` will be so useful. In each class I explain the awkward process of either rotating before you draw the shape (how it works in Processing.. the rotate after in a loop is 'getting lucky' and clever) – or after you draw the shape, having to use the `obj.transformPropName = ...` technique.. then it's a weird `function()` vs `=` syntax. This `b.transform()` would solve that.

I had also made a note (forget if an issue has been created yet?) to easily set the anchorPoint before a rotation:

Proposal:
`b.referencePoint()` or `b.anchorPoint()`
- which seems to not be specific to any single item, but general preference for future transformations..
- OR, it's simply also part of the b.transform().. but it's more tricky/specific than the other transform properties, so should likely get it's own function.

`app.activeWindow.transformReferencePoint = AnchorPoint.CENTER_ANCHOR;`

https://indisnip.wordpress.com/2010/07/24/rotate-single-selected-object/
http://jongware.mit.edu/idcs5js/pe_AnchorPoint.html

**Comment by trych** [comment]
> I had also made a note (forget if an issue has been created yet?) to easily set the anchorPoint before a rotation

Yes, I also thought recently that we need such a function. There is no issue yet, I'll open one, let's move the discussion there, to leave this issue for the matrix topic only.

**Comment by bohnacker** [comment]
I just created a pull request with the changes I made to transformation.js and shape.js. I think, this solves the problem that the matrix transformation behaved differently than in Processing. 

I try to explain what I did:
I found that the transformations already worked quite correct when an object was created at the origin. Something like
```
b.rotate( b.radians(45) );
b.translate( 100, 50 );
b.rect(0, 0, 20, 20);
```
produced already the same result as Processing would do.

The problem starts when x or y is not 0. Let's have this example:
```
b.rotate( b.radians(45) );
b.translate( 100, 50 );
b.rect(40, 10, 20, 20);
```
The `rotate` and `translate` calls build a transformation matrix M. In order to have the correct effect when applying this matrix to the object we need to pre-apply a translation to move the object to the origin and move it back to its position afterwards. Expressed with matrix multiplication its just something like this:
`T(-x, -y) * M * T(x, y)`

The result is a matrix that cancels out the special way that InDesign does its transformations. At least I hope so.
